### PR TITLE
BUILD-278: Check cgroup v1 and cgroup v2 Files for Quota

### PR DESCRIPTION
**Pull request by gabemontero**
/assign @adambkaplan 
/assign @nalind 

there was no existing unit testing around `GetCGroupLimits()` or `readIn64` that I saw

the simplicity of the fix did not motivate me to take that on, but of course am curious on your respective takes

### CODE REVIEW

File: pkg/build/builder/util_linux.go
Comment (nalind) [inline pkg/build/builder/util_linux.go:19]:
```suggestion
		"/sys/fs/cgroup/memory.max",
```

File: pkg/build/builder/util_linux.go
Comment (nalind) [inline pkg/build/builder/util_linux.go:18]: A quick spin through the crun and runc sources suggests that neither of them set `memory.high`, so we probably want to skip it.

File: pkg/build/builder/util_linux.go
Comment (nalind) [inline pkg/build/builder/util_linux.go:6]: Probably want to drop this hunk.

File: pkg/build/builder/util_linux.go
Comment (gabemontero) [reply pkg/build/builder/util_linux.go]: update thanks

File: pkg/build/builder/util_linux.go
Comment (gabemontero) [reply pkg/build/builder/util_linux.go:5]: moved thanks

File: pkg/build/builder/util_linux.go
Comment (gabemontero) [reply pkg/build/builder/util_linux.go:18]: ok, which lines up with Guiseppe just suggesting `memory.max` in the bugzilla

**Comment by gabemontero** [comment]
I think that is all your comments @nalind thanks

File: pkg/build/builder/dockerutil.go
Comment (adambkaplan) [inline pkg/build/builder/dockerutil.go:129]: Need a unit test for this function

**Review by adambkaplan — CHANGES_REQUESTED** [review]

File: pkg/build/builder/dockerutil.go
Comment (gabemontero) [reply pkg/build/builder/dockerutil.go:129]: so per my comment in https://github.com/openshift/builder/pull/246#issue-655174072 it would make sense to add unit tests not just for this function, but some of the existing code which does not have unit tests (i.e. `readInt64`)

so you are officially asking for that @adambkaplan ?

Comment (adambkaplan) [reply pkg/build/builder/dockerutil.go:129]: Yes - there is a general ask across OpenShift to improve test code coverage. Adding scoped unit tests to new functions (and in this case, new functions that extend existing functions) moves the needle in the right direction.

Comment (gabemontero) [reply pkg/build/builder/dockerutil.go:129]: unit test added @adambkaplan

Comment (gabemontero) [reply pkg/build/builder/dockerutil.go:129]: actually sit tight @adambkaplan 

per https://prow.ci.openshift.org/view/gs/origin-ci-test/pr-logs/pull/openshift_builder/246/pull-ci-openshift-builder-master-unit/1397983052998643712#1:build-log.txt%3A12

may have a golang version discrepancy to sort out .... `os.WriteFile` works for me locally but is noted as undefined in CI

Comment (gabemontero) [reply pkg/build/builder/dockerutil.go:129]: sure  enough, things moved between golang 1.15 and 1.16 per the overview in https://golang.org/pkg/io/ioutil/ and openshift/builder is still at 1.15 per https://github.com/openshift/release/blob/master/ci-operator/config/openshift/builder/openshift-builder-master.yaml#L37

I know you have limited availability this afternoon @adambkaplan so I'm for now going to assume we do not want to bump golang for openshift/builder this late in the cycle, and I'll switch from os.WriteFile to io/ioutil.WriteFile

Comment (gabemontero) [reply pkg/build/builder/dockerutil.go:129]: ok unit tests now clean @adambkaplan

File: pkg/build/builder/dockerutil.go
Comment (kolyshkin) [inline pkg/build/builder/dockerutil.go:138]: So basically you're reading the file twice -- this is redundant.

I suggest to separate out the portion of readInt64 that parses the value and re-use it from here.

Something like this maybe

```diff
diff --git a/pkg/build/builder/dockerutil.go b/pkg/build/builder/dockerutil.go
index 764e06655..1ef30681a 100644
--- a/pkg/build/builder/dockerutil.go
+++ b/pkg/build/builder/dockerutil.go
@@ -111,7 +111,24 @@ func readInt64(filePath string) (int64, error) {
        if err != nil {
                return -1, err
        }
+
+       return parseInt64(strings.TrimSpace(string(data)))
+}
+
+func readMaxStringOrInt64(filePath string) (int64, error) {
+       data, err := ioutil.ReadFile(filePath)
+       if err != nil {
+               return -1, err
+       }
        s := strings.TrimSpace(string(data))
+       if s == "max" {
+               return math.MaxInt64, nil
+       }
+
+       return parseInt64(s)
+}
+
+func parseInt64(s string) (int64, error) {
        val, err := strconv.ParseInt(s, 10, 64)
        // overflow errors are ok, we'll get return a math.MaxInt64 value which is more
        // than enough anyway.  For underflow we'll return MinInt64 and the error.
```

File: pkg/build/builder/dockerutil_test.go
Comment (kolyshkin) [inline pkg/build/builder/dockerutil_test.go:227]: No need for this as `ioutil.WriteFile` creates the file.

File: pkg/build/builder/dockerutil_test.go
Comment (kolyshkin) [inline pkg/build/builder/dockerutil_test.go:228]: `0777` looks a bit excessive. Perhaps
```suggestion
	err = ioutil.WriteFile(fileName, []byte(fileContent), 0o644)
```

File: pkg/build/builder/util_linux.go
Comment (kolyshkin) [inline pkg/build/builder/util_linux.go:18]: Formatting nit:
```suggestion
	filesToTry := []string{
		"/sys/fs/cgroup/memory/memory.limit_in_bytes",
```

File: pkg/build/builder/util_linux.go
Comment (kolyshkin) [inline pkg/build/builder/util_linux.go:23]: Trying to read cgroup v2 file on a cgroup v1 system, or vice versa, is not very efficient.

Trying to read cgroup v2 file on _any_ error from reading cgroup v1 file is also not good.

I suggest using something like IsCgroupV2UnifiedMode() from https://github.com/opencontainers/runc/blob/960b5fdb5acde1063960a80eb224672faf244e02/libcontainer/cgroups/utils.go#L34

and do this:
```go
file := "/sys/fs/cgroup/memory/memory.limit_in_bytes"
if IsCgroupV2UnifiedMode() {
   file = "/sys/fs/cgroup/memory.max"
}
byteLimit, err := readMaxStringOrInt64(file)
....
```

**Review by kolyshkin — CHANGES_REQUESTED** [review]

**Comment by gabemontero** [comment]
updates pushed @kolyshkin @nalind ptal thanks

File: pkg/build/builder/dockerutil.go
Comment (kolyshkin) [inline pkg/build/builder/dockerutil.go:134]: nit: extra whitespace

Comment (kolyshkin) [reply pkg/build/builder/dockerutil.go:134]: Funny there's an (very old) open issue for gofmt to address this particular case (https://github.com/golang/go/issues/6996)

**Review by kolyshkin — COMMENTED** [review]
a whitespace nit, otherwise LGTM

**Review by adambkaplan — APPROVED** [review]
/approve

Outstanding format nit, otherwise looks good.

**Comment by gabemontero** [comment]
white space removed @adambkaplan @kolyshkin thanks folks

File: pkg/build/builder/util_linux.go
Comment (kolyshkin) [inline pkg/build/builder/util_linux.go:17]: Technically this comment belongs to cgroupv2 case inside the `if` block, but that's not important.

In the future, though, we might need to also check `memory.high` (in case `memory.max` returns `max`.

**Review by kolyshkin — APPROVED** [review]
LGTM

File: pkg/build/builder/util_linux.go
Comment (kolyshkin) [inline pkg/build/builder/util_linux.go:20]: Sorry I haven't raised this concern earlier, but here it is.

cgroup v2 does not have "memory.max" file in the top-level cgroup **on the host**, so this won't work **on the host** (unless you add a fallback to assume the limit is `max` or somesuch).

If this is done inside a container, this check will work, as the container has its own cgroup mounted to /sys/fs/cgroup, not the top one.

Hope that helps.

Comment (gabemontero) [reply pkg/build/builder/util_linux.go:20]: thanks @kolyshkin - that clarification, and the additional host level verification details you added in the associated bugzilla, should help with the confusion @wewang58 was having in trying to verify a proper set up to verify.

As it turns out, this logic is in fact running in a container running in the build pod.

@adambkaplan I think we are ready for the "official" lgtm so we can get this merged, unless you had any other questions.

**Comment by gabemontero** [comment]
per @rphillips cgroupsv2 support is out wrt openshift and it is "too soon" to drop this in.... QE has now "user acceptable" form for installing openshift with cgroupsv2

I'm going to

/hold 

this vs. closing for now to make it easier to find and resuscitate this when in fact cgroupsv2 gets officially into openshift plan pipeline

thanks again @kolyshkin for you assistance

**Comment by adambkaplan** [comment]
/retitle BUILD-278: Check cgroup v1 and cgroup v2 Files for Quota

**Comment by gabemontero** [comment]
@vrutkovs mentioned in bz 1949438 that at the moment cgroupsv2 is tech preview for 4.9

File: pkg/build/builder/util_linux.go
Comment (kolyshkin) [reply pkg/build/builder/util_linux.go:20]: It's all good then.

**Review by kolyshkin — APPROVED** [review]
LGTM

**Comment by vrutkovs** [comment]
>CI is not ready for it yet 

CI job to test this is added in https://github.com/openshift/release/pull/19115, I suppose this is good to go in 4.9 now?

File: pkg/build/builder/util_linux.go
Comment (giuseppe) [reply pkg/build/builder/util_linux.go]: for completeness, both runc/crun supports setting `memory.high` through the unified map (podman exposes it with `--cgroup-conf`) but I don't think we should worry about it, as a cgroup is still allowed to allocate more memory until it reaches `memory.max`

**Review by giuseppe — APPROVED** [review]
LGTM

**Review by adambkaplan — APPROVED** [review]
/lgtm

**Comment by adambkaplan** [comment]
/hold cancel

I imagine we need to merge this for the cgroupsv2 jobs to go green.

**Comment by gabemontero** [comment]
> /hold cancel
> 
> I imagine we need to merge this for the cgroupsv2 jobs to go green.

last I heard @adambkaplan about this was https://github.com/openshift/builder/pull/246#issuecomment-853860972 and a ping from @rphillips where he said CI was not ready for cgroupsv2

just double checking are you aware with master branch opening up for 4.9 where the install changes needed for CI (and test for that matter) to get this are in place?

**Comment by adambkaplan** [comment]
> just double checking are you aware with master branch opening up for 4.9 where the install changes needed for CI (and test for that matter) to get this are in place?

Yes - ack that we won't have CI running this end to end until we can install with cgroupsv2.

**Comment by gabemontero** [comment]
Hey @kolyshkin @nalind : so @vrutkovs is trying some cgroupv2 CI rehearsals over with PR 

In that PR, where presumably cgroupv2 is set up correctly, I can confirm in the logs that our us our use of the openshicontainers/runc API is kicking in at 

https://github.com/openshift/builder/blob/814bbb090228374742e31889c7fe3df0c3662c7a/pkg/build/builder/util_linux.go#L19-L21

because I see the code picking `/sys/fs/cgroup/memory.max` as the file to access.

However, we are ultimately getting error logs like 

`2021-07-12T09:05:29.940355077Z error: failed to retrieve cgroup limits: cannot determine cgroup limits: open /sys/fs/cgroup/memory.max: no such file or directory`

Did we get the name of the file wrong?
We currently do not assume the file is there with https://github.com/openshift/builder/blob/814bbb090228374742e31889c7fe3df0c3662c7a/pkg/build/builder/util_linux.go#L26-L28 but perhaps seeing if `/sys/fs/cgroup` not existing is an insufficient check for ignoring the lack of `/sys/fs/cgroup/memory.max` ?

Or do you see something in the test setup @vrutkovs has ?

I could start adding (short term/temporary) debug in that path to display the full contents of `/sys/fs.cgroup` but I decided to pull you guys in first and get any possibly quick sanity checks from you all.

@adambkaplan fyi

thanks

**Comment by gabemontero** [comment]
edited my prior comment with the link to @vrutkovs 's test enablement PR

**Comment by vrutkovs** [comment]
Here's what I see on the node:
```
$ oc debug node/ip-10-0-140-154.ec2.internal                                                      
Starting pod/ip-10-0-140-154ec2internal-debug ...
To use host binaries, run `chroot /host`
chroot /host
Pod IP: 10.0.140.154
If you don't see a command prompt, try pressing enter.
sh-4.4# chroot /host
sh-4.4# cat /proc/cmdline 
BOOT_IMAGE=(hd0,gpt3)/ostree/rhcos-20246fb86657495594dfd90b85bcfa58f8a400916e1580611a5b7ea95e77f783/vmlinuz-4.18.0-305.7.1.el8_4.x86_64 random.trust_cpu=on console=tty0 console=ttyS0,115200n8 ostree=/ostree/boot.0/rhcos/20246fb86657495594dfd90b85bcfa58f8a400916e1580611a5b7ea95e77f783/0 ignition.platform.id=aws root=UUID=12afac90-6b11-458a-9f0b-1f5303f2b1f9 rw rootflags=prjquota systemd.unified_cgroup_hierarchy=1
sh-4.4# ls -la /sys/fs/cgroup/
total 0
dr-xr-xr-x.  6 root root 0 Jul 12 15:51 .
drwxr-xr-x.  9 root root 0 Jul 12 15:50 ..
-r--r--r--.  1 root root 0 Jul 12 15:50 cgroup.controllers
-rw-r--r--.  1 root root 0 Jul 12 15:51 cgroup.max.depth
-rw-r--r--.  1 root root 0 Jul 12 15:51 cgroup.max.descendants
-rw-r--r--.  1 root root 0 Jul 12 15:50 cgroup.procs
-r--r--r--.  1 root root 0 Jul 12 15:51 cgroup.stat
-rw-r--r--.  1 root root 0 Jul 12 16:12 cgroup.subtree_control
-rw-r--r--.  1 root root 0 Jul 12 15:51 cgroup.threads
-rw-r--r--.  1 root root 0 Jul 12 15:51 cpu.pressure
-r--r--r--.  1 root root 0 Jul 12 15:51 cpuset.cpus.effective
-r--r--r--.  1 root root 0 Jul 12 15:51 cpuset.mems.effective
drwxr-xr-x.  2 root root 0 Jul 12 15:50 init.scope
-rw-r--r--.  1 root root 0 Jul 12 15:51 io.pressure
-r--r--r--.  1 root root 0 Jul 12 15:51 io.stat
drwxr-xr-x.  4 root root 0 Jul 12 15:51 kubepods.slice
-rw-r--r--.  1 root root 0 Jul 12 15:51 memory.pressure
-r--r--r--.  1 root root 0 Jul 12 15:51 memory.stat
drwxr-xr-x. 77 root root 0 Jul 12 15:50 system.slice
drwxr-xr-x.  2 root root 0 Jul 12 15:51 user.slice
```

And in the container:
```
$ oc run fedora --image=quay.io/fedora/fedora:34 --command -- sleep infinity
pod/fedora created
$ oc exec -ti fedora -- sh
sh-5.0# ls -la /sys/fs/cgroup/
total 0
drwxr-xr-x.  2 root root 0 Jul 12 16:15 .
drwxr-xr-x. 10 root root 0 Jul 12 16:15 ..
-r--r--r--.  1 root root 0 Jul 12 16:15 cgroup.controllers
-r--r--r--.  1 root root 0 Jul 12 16:15 cgroup.events
-rw-r--r--.  1 root root 0 Jul 12 16:15 cgroup.freeze
-rw-r--r--.  1 root root 0 Jul 12 16:15 cgroup.max.depth
-rw-r--r--.  1 root root 0 Jul 12 16:15 cgroup.max.descendants
-rw-r--r--.  1 root root 0 Jul 12 16:15 cgroup.procs
-r--r--r--.  1 root root 0 Jul 12 16:15 cgroup.stat
-rw-r--r--.  1 root root 0 Jul 12 16:15 cgroup.subtree_control
-rw-r--r--.  1 root root 0 Jul 12 16:15 cgroup.threads
-rw-r--r--.  1 root root 0 Jul 12 16:15 cgroup.type
-rw-r--r--.  1 root root 0 Jul 12 16:15 cpu.max
-rw-r--r--.  1 root root 0 Jul 12 16:15 cpu.pressure
-r--r--r--.  1 root root 0 Jul 12 16:15 cpu.stat
-rw-r--r--.  1 root root 0 Jul 12 16:15 cpu.weight
-rw-r--r--.  1 root root 0 Jul 12 16:15 cpu.weight.nice
-rw-r--r--.  1 root root 0 Jul 12 16:15 cpuset.cpus
-r--r--r--.  1 root root 0 Jul 12 16:15 cpuset.cpus.effective
-rw-r--r--.  1 root root 0 Jul 12 16:15 cpuset.cpus.partition
-rw-r--r--.  1 root root 0 Jul 12 16:15 cpuset.mems
-r--r--r--.  1 root root 0 Jul 12 16:15 cpuset.mems.effective
-r--r--r--.  1 root root 0 Jul 12 16:15 hugetlb.1GB.current
-r--r--r--.  1 root root 0 Jul 12 16:15 hugetlb.1GB.events
-r--r--r--.  1 root root 0 Jul 12 16:15 hugetlb.1GB.events.local
-rw-r--r--.  1 root root 0 Jul 12 16:15 hugetlb.1GB.max
-r--r--r--.  1 root root 0 Jul 12 16:15 hugetlb.2MB.current
-r--r--r--.  1 root root 0 Jul 12 16:15 hugetlb.2MB.events
-r--r--r--.  1 root root 0 Jul 12 16:15 hugetlb.2MB.events.local
-rw-r--r--.  1 root root 0 Jul 12 16:15 hugetlb.2MB.max
-rw-r--r--.  1 root root 0 Jul 12 16:15 io.bfq.weight
-rw-r--r--.  1 root root 0 Jul 12 16:15 io.latency
-rw-r--r--.  1 root root 0 Jul 12 16:15 io.max
-rw-r--r--.  1 root root 0 Jul 12 16:15 io.pressure
-r--r--r--.  1 root root 0 Jul 12 16:15 io.stat
-r--r--r--.  1 root root 0 Jul 12 16:15 memory.current
-r--r--r--.  1 root root 0 Jul 12 16:15 memory.events
-rw-r--r--.  1 root root 0 Jul 12 16:15 memory.high
-rw-r--r--.  1 root root 0 Jul 12 16:15 memory.low
-rw-r--r--.  1 root root 0 Jul 12 16:15 memory.max
-rw-r--r--.  1 root root 0 Jul 12 16:15 memory.min
-rw-r--r--.  1 root root 0 Jul 12 16:15 memory.oom.group
-rw-r--r--.  1 root root 0 Jul 12 16:15 memory.pressure
-r--r--r--.  1 root root 0 Jul 12 16:15 memory.stat
-r--r--r--.  1 root root 0 Jul 12 16:15 memory.swap.current
-r--r--r--.  1 root root 0 Jul 12 16:15 memory.swap.events
-rw-r--r--.  1 root root 0 Jul 12 16:15 memory.swap.high
-rw-r--r--.  1 root root 0 Jul 12 16:15 memory.swap.max
-r--r--r--.  1 root root 0 Jul 12 16:15 pids.current
-r--r--r--.  1 root root 0 Jul 12 16:15 pids.events
-rw-r--r--.  1 root root 0 Jul 12 16:15 pids.max
-r--r--r--.  1 root root 0 Jul 12 16:15 rdma.current
-rw-r--r--.  1 root root 0 Jul 12 16:15 rdma.max
```

**Comment by gabemontero** [comment]
Thanks @vrutkovs - certainly helpful

So the current call for cgroup stuff is in the build **pod**, when it is prepping the build container, which is potentially different than both your node snapshot (where `memory.max` is missing) and your direct fedora container invocation snaphsot (where `memory.max` is present).

Certainly in both, `/sys/fs/cgroup/` exists.

The reasonable assumption based on the behavior exhibited would be that the build pod looks like the node.  A dir structure like the node would result in the behavior exhibited.

My sense is that we should still verify the contents while in the build pod.  But assuming it is like the node, then get agreement when `/sys/fs/cgroup` exists, but `memory.max` does not, we assume that this the "default" case.

But I do not consider myself an SME here and would like some consensus / confirmation from @kolyshkin @nalind @adambkaplan

**Comment by gabemontero** [comment]
I've created https://github.com/openshift/builder/pull/252 to potentially address the results of @vrutkovs 's CI testing attempt and move the associated discussion to a open vs. merged PR

**Pull request merged**
